File: Dart/15-Greedy/Algorithms/3-Kruskal’sAlgorithm(MST)/Kruskal’sAlgorithm(MST)Pro.dart
import 'dart:collection';
import 'package:collection/collection.dart'; // Import for HeapPriorityQueue

// Class to represent an edge in the graph
class Edge {
  int source;
  int destination;
  int weight;
  String? type; // Added for real-world examples

  Edge(this.source, this.destination, this.weight, {this.type});

  @override
  String toString() {
    return 'Edge{source: $source, destination: $destination, weight: $weight, type: $type}';
  }
}

// Function to find the parent of a node (with path compression)
int find(List<int> parent, int i) {
  if (parent[i] == i) return i;
  return parent[i] = find(parent, parent[i]); // Path compression
}

// Function to perform union of two sets (using rank)
void union(List<int> parent, List<int> rank, int x, int y) {
  int xRoot = find(parent, x);
  int yRoot = find(parent, y);

  if (rank[xRoot] < rank[yRoot]) {
    parent[xRoot] = yRoot;
  } else if (rank[xRoot] > rank[yRoot]) {
    parent[yRoot] = xRoot;
  } else {
    parent[yRoot] = xRoot;
    rank[xRoot]++;
  }
}

// 1. Basic Kruskal's Algorithm
List<Edge> kruskalMST(List<Edge> edges, int numVertices) {
  edges.sort((a, b) => a.weight.compareTo(b.weight));
  List<int> parent = List.generate(numVertices, (i) => i);
  List<int> rank = List.generate(numVertices, (i) => 0);
  List<Edge> minimumSpanningTree = [];
  int edgeCount = 0;

  for (Edge edge in edges) {
    int sourceParent = find(parent, edge.source);
    int destinationParent = find(parent, edge.destination);
    if (sourceParent != destinationParent) {
      minimumSpanningTree.add(edge);
      union(parent, rank, sourceParent, destinationParent);
      edgeCount++;
    }
    if (edgeCount == numVertices - 1) break;
  }
  return minimumSpanningTree;
}

// 2. Kruskal's Algorithm using a Priority Queue
List<Edge> kruskalMSTPriorityQueue(List<Edge> edges, int numVertices) {
  final PriorityQueue<Edge> pq = HeapPriorityQueue((a, b) => a.weight.compareTo(b.weight));
  pq.addAll(edges);

  List<int> parent = List.generate(numVertices, (i) => i);
  List<int> rank = List.generate(numVertices, (i) => 0);
  List<Edge> minimumSpanningTree = [];
  int edgeCount = 0;

  while (pq.isNotEmpty) {
    Edge edge = pq.removeFirst();
    int sourceParent = find(parent, edge.source);
    int destinationParent = find(parent, edge.destination);
    if (sourceParent != destinationParent) {
      minimumSpanningTree.add(edge);
      union(parent, rank, sourceParent, destinationParent);
      edgeCount++;
    }
    if (edgeCount == numVertices - 1) break;
  }
  return minimumSpanningTree;
}

// 3. Kruskal's Algorithm with Edge Types (Real-World Application)
List<Edge> kruskalMSTWithType(List<Edge> edges, int numVertices,
    {String requiredType = 'all'}) {
  edges.sort((a, b) => a.weight.compareTo(b.weight));
  List<int> parent = List.generate(numVertices, (i) => i);
  List<int> rank = List.generate(numVertices, (i) => 0);
  List<Edge> minimumSpanningTree = [];
  int edgeCount = 0;

  for (Edge edge in edges) {
    int sourceParent = find(parent, edge.source);
    int destinationParent = find(parent, edge.destination);
    // Filter by edge type if required
    if (sourceParent != destinationParent &&
        (requiredType == 'all' || edge.type == requiredType)) {
      minimumSpanningTree.add(edge);
      union(parent, rank, sourceParent, destinationParent);
      edgeCount++;
    }
    if (edgeCount == numVertices - 1) break;
  }
  return minimumSpanningTree;
}

// 4. Kruskal's Algorithm with Capacity Constraints (Real-World)
List<Edge> kruskalMSTWithCapacity(List<Edge> edges, int numVertices,
    Map<int, int> capacity) {
  edges.sort((a, b) => a.weight.compareTo(b.weight));
  List<int> parent = List.generate(numVertices, (i) => i);
  List<int> rank = List.generate(numVertices, (i) => 0);
  List<Edge> minimumSpanningTree = [];
  int edgeCount = 0;
  Map<int, int> currentLoad = {}; // Track load on each node

  for (Edge edge in edges) {
    int sourceParent = find(parent, edge.source);
    int destinationParent = find(parent, edge.destination);

    if (sourceParent != destinationParent) {
      // Check capacity constraints
      int sourceLoad = currentLoad[edge.source] ?? 0;
      int destinationLoad = currentLoad[edge.destination] ?? 0;
      if (sourceLoad + edge.weight <= (capacity[edge.source] ?? 100000) && //Assume 100000 is max capacity if not provided.
          destinationLoad + edge.weight <= (capacity[edge.destination] ?? 100000)) {
        minimumSpanningTree.add(edge);
        union(parent, rank, sourceParent, destinationParent);
        edgeCount++;
        // Update current loads
        currentLoad[edge.source] = sourceLoad + edge.weight;
        currentLoad[edge.destination] = destinationLoad + edge.weight;
      }
    }
    if (edgeCount == numVertices - 1) break;
  }
  return minimumSpanningTree;
}

void main() {
  // Example graph:
  // (0)--(1)--(2)
  //  |    |    |
  // (3)--(4)--(5)
  List<Edge> edges = [
    Edge(0, 1, 4, type: 'cable'),
    Edge(0, 3, 8, type: 'pipe'),
    Edge(1, 2, 9, type: 'cable'),
    Edge(1, 4, 2, type: 'cable'),
    Edge(2, 5, 5, type: 'cable'),
    Edge(3, 4, 3, type: 'pipe'),
    Edge(3, 6, 6, type: 'pipe'),
    Edge(4, 5, 7, type: 'cable'),
    Edge(4, 6, 1, type: 'pipe'),
    Edge(5, 6, 8, type: 'cable'),
  ];
  int numVertices = 7;

  // 1. Basic Kruskal's
  print("1. Basic Kruskal's MST:");
  List<Edge> mst1 = kruskalMST(edges, numVertices);
  mst1.forEach(print);

  // 2. Kruskal's with Priority Queue
  print("\n2. Kruskal's MST with Priority Queue:");
  List<Edge> mst2 = kruskalMSTPriorityQueue(edges, numVertices);
  mst2.forEach(print);

  // 3. Kruskal's with Edge Types
  print("\n3. Kruskal's MST with Cable Edges:");
  List<Edge> mst3 = kruskalMSTWithType(edges, numVertices, requiredType: 'cable');
  mst3.forEach(print);

  // 4. Kruskal's with Capacity Constraints
  print("\n4. Kruskal's MST with Capacity Constraints:");
  Map<int, int> capacity = {
    0: 10,
    1: 15,
    2: 20,
    3: 12,
    4: 18,
    5: 25,
    6: 30
  }; // Example capacities
  List<Edge> mst4 = kruskalMSTWithCapacity(edges, numVertices, capacity);
  mst4.forEach(print);
}

File: Java/15-Greedy/Algorithms/3-Kruskal’sAlgorithm(MST)/Kruskal’sAlgorithm(MST)Pro.java
import java.util.*;
import java.io.*;

public class KruskalMST {

    // Basic Edge class to represent an edge in the graph
    static class Edge implements Comparable<Edge> {
        int src, dest, weight;

        public Edge(int src, int dest, int weight) {
            this.src = src;
            this.dest = dest;
            this.weight = weight;
        }

        @Override
        public int compareTo(Edge other) {
            return this.weight - other.weight; // Sort edges by weight
        }

        @Override
        public String toString() {
            return "(" + src + " -- " + dest + ") : " + weight;
        }
    }

    // 1. Kruskal's Algorithm using Union-Find (Disjoint Set) - Classic Implementation
    //    - Uses an array to represent the parent of each node in the disjoint set.
    //    - find() uses path compression for optimization.
    //    - union() uses union by rank for optimization.
    public static List<Edge> kruskalMST_UnionFind(int numVertices, List<Edge> edges) {
        List<Edge> mst = new ArrayList<>();
        Collections.sort(edges); // Sort edges by weight

        int[] parent = new int[numVertices];
        int[] rank = new int[numVertices]; // Added for union by rank
        for (int i = 0; i < numVertices; i++) {
            parent[i] = i; // Initialize each node as its own parent
            rank[i] = 0;  // Initialize rank of each node to 0
        }

        for (Edge edge : edges) {
            int rootSrc = find(parent, edge.src);
            int rootDest = find(parent, edge.dest);

            if (rootSrc != rootDest) {
                mst.add(edge);
                union(parent, rank, rootSrc, rootDest); // Use union by rank
            }
        }
        return mst;
    }

    // Find operation with path compression
    private static int find(int[] parent, int i) {
        if (parent[i] == i)
            return i;
        return parent[i] = find(parent, parent[i]); // Path compression
    }

    // Union operation using union by rank
    private static void union(int[] parent, int[] rank, int x, int y) {
        int rootX = find(parent, x);
        int rootY = find(parent, y);

        if (rootX != rootY) {
            if (rank[rootX] < rank[rootY]) {
                parent[rootX] = rootY;
            } else if (rank[rootX] > rank[rootY]) {
                parent[rootY] = rootX;
            } else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
        }
    }

    // 2. Kruskal's Algorithm using a HashMap for Disjoint Set - Alternative Representation
    //    - Uses a HashMap to store the parent of each node.  Useful if node IDs are not simple integers.
    public static List<Edge> kruskalMST_HashMap(Set<Integer> vertices, List<Edge> edges) {
        List<Edge> mst = new ArrayList<>();
        Collections.sort(edges);

        Map<Integer, Integer> parentMap = new HashMap<>();
        Map<Integer, Integer> rankMap = new HashMap<>(); // Added for union by rank
        for (Integer vertex : vertices) {
            parentMap.put(vertex, vertex);
            rankMap.put(vertex, 0);
        }

        for (Edge edge : edges) {
            int rootSrc = findMap(parentMap, edge.src);
            int rootDest = findMap(parentMap, edge.dest);

            if (rootSrc != rootDest) {
                mst.add(edge);
                unionMap(parentMap, rankMap, rootSrc, rootDest);
            }
        }
        return mst;
    }

    // Find operation for HashMap
    private static int findMap(Map<Integer, Integer> parentMap, int i) {
        if (parentMap.get(i) == i)
            return i;
        int root = findMap(parentMap, parentMap.get(i));
        parentMap.put(i, root); // Path compression
        return root;
    }

    // Union operation for HashMap using union by rank
    private static void unionMap(Map<Integer, Integer> parentMap, Map<Integer, Integer> rankMap, int x, int y) {
        int rootX = findMap(parentMap, x);
        int rootY = findMap(parentMap, y);

        if (rootX != rootY) {
            if (rankMap.get(rootX) < rankMap.get(rootY)) {
                parentMap.put(rootX, rootY);
            } else if (rankMap.get(rootX) > rankMap.get(rootY)) {
                parentMap.put(rootY, rootX);
            } else {
                parentMap.put(rootY, rootX);
                rankMap.put(rootX, rankMap.get(rootX) + 1);
            }
        }
    }

    // 3. Kruskal's Algorithm with Path Compression and Rank Optimization (Optimized)
    //    - This is a direct optimization of the first approach, making it more efficient.
    public static List<Edge> kruskalMST_Optimized(int numVertices, List<Edge> edges) {
        List<Edge> mst = new ArrayList<>();
        Collections.sort(edges);

        int[] parent = new int[numVertices];
        int[] rank = new int[numVertices];
        for (int i = 0; i < numVertices; i++) {
            parent[i] = i;
            rank[i] = 0;
        }

        for (Edge edge : edges) {
            int rootSrc = findOptimized(parent, edge.src);
            int rootDest = findOptimized(parent, edge.dest);

            if (rootSrc != rootDest) {
                mst.add(edge);
                unionOptimized(parent, rank, rootSrc, rootDest);
            }
        }
        return mst;
    }

    // Optimized find with path compression
    private static int findOptimized(int[] parent, int i) {
        if (parent[i] != i)
            parent[i] = findOptimized(parent, parent[i]);
        return parent[i];
    }

    // Optimized union by rank
    private static void unionOptimized(int[] parent, int[] rank, int x, int y) {
        int rootX = findOptimized(parent, x);
        int rootY = findOptimized(parent, y);
        if (rootX != rootY) {
            if (rank[rootX] < rank[rootY])
                parent[rootX] = rootY;
            else if (rank[rootX] > rank[rootY])
                parent[rootY] = rootX;
            else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
        }
    }

    // 4. Kruskal's Algorithm for Disconnected Graphs
    //    - Handles graphs that may have multiple disconnected components.
    public static List<Edge> kruskalMST_Disconnected(int numVertices, List<Edge> edges) {
        List<Edge> mst = new ArrayList<>();
        Collections.sort(edges);

        int[] parent = new int[numVertices];
        int[] rank = new int[numVertices];
        for (int i = 0; i < numVertices; i++) {
            parent[i] = i;
            rank[i] = 0;
        }

        int numTrees = numVertices; // Initially, each vertex is its own tree
        for (Edge edge : edges) {
            int rootSrc = findOptimized(parent, edge.src);
            int rootDest = findOptimized(parent, edge.dest);

            if (rootSrc != rootDest) {
                mst.add(edge);
                unionOptimized(parent, rank, rootSrc, rootDest);
                numTrees--; // Joining two trees reduces the number of trees by 1
            }
            if (numTrees == 1) break; // Optimization: Stop when only one tree remains
        }
        if(numTrees > 1){
            System.out.println("Graph is disconnected.  MST is a Minimum Spanning Forest.");
        }
        return mst;
    }


    // 5. Kruskal's Algorithm with Edge Class Enhancement (Real-World Application)
    //    -  Demonstrates a more complex Edge class that might be used in a real-world scenario,
    //       such as network design, where edges have additional properties like bandwidth and cost.
    //    -  The `compareTo` method is adjusted to prioritize bandwidth and then cost (for example).
    static class NetworkEdge implements Comparable<NetworkEdge> {
        int src, dest;
        double bandwidth; // Example: Bandwidth of the connection
        double cost;     // Example: Cost of the connection

        public NetworkEdge(int src, int dest, double bandwidth, double cost) {
            this.src = src;
            this.dest = dest;
            this.bandwidth = bandwidth;
            this.cost = cost;
        }

        @Override
        public int compareTo(NetworkEdge other) {
            // Prioritize higher bandwidth, then lower cost
            if (this.bandwidth != other.bandwidth) {
                return Double.compare(other.bandwidth, this.bandwidth); // Descending order of bandwidth
            }
            return Double.compare(this.cost, other.cost); // Ascending order of cost
        }

        @Override
        public String toString() {
            return "(" + src + " -- " + dest + ") : Bandwidth=" + bandwidth + ", Cost=" + cost;
        }
    }

    public static List<NetworkEdge> kruskalMST_Network(int numVertices, List<NetworkEdge> networkEdges) {
        List<NetworkEdge> mst = new ArrayList<>();
        Collections.sort(networkEdges);

        int[] parent = new int[numVertices];
        int[] rank = new int[numVertices];
        for (int i = 0; i < numVertices; i++) {
            parent[i] = i;
            rank[i] = 0;
        }

        for (NetworkEdge edge : networkEdges) {
            int rootSrc = findOptimized(parent, edge.src);
            int rootDest = findOptimized(parent, edge.dest);

            if (rootSrc != rootDest) {
                mst.add(edge);
                unionOptimized(parent, rank, rootSrc, rootDest);
            }
        }
        return mst;
    }



    public static void main(String[] args) {
        // Example graph represented as a list of edges
        int numVertices = 7;
        List<Edge> edges = new ArrayList<>();
        edges.add(new Edge(0, 1, 7));
        edges.add(new Edge(0, 2, 8));
        edges.add(new Edge(1, 2, 3));
        edges.add(new Edge(1, 3, 6));
        edges.add(new Edge(2, 3, 4));
        edges.add(new Edge(2, 4, 9));
        edges.add(new Edge(3, 4, 5));
        edges.add(new Edge(3, 5, 10));
        edges.add(new Edge(4, 5, 8));
        edges.add(new Edge(4, 6, 11));
        edges.add(new Edge(5, 6, 9));

        System.out.println("Example Graph Edges:");
        for (Edge edge : edges) {
            System.out.println(edge);
        }
        System.out.println("\n--------------------------------------------------------");

        // 1. Kruskal's Algorithm using Union-Find
        System.out.println("Kruskal's MST (Union-Find):");
        List<Edge> mst1 = kruskalMST_UnionFind(numVertices, new ArrayList<>(edges)); // Pass a copy to avoid modifying the original
        for (Edge edge : mst1) {
            System.out.println(edge);
        }
        System.out.println("\n--------------------------------------------------------");

        // 2. Kruskal's Algorithm using HashMap
        Set<Integer> verticesSet = new HashSet<>();
        for (Edge edge : edges) {
            verticesSet.add(edge.src);
            verticesSet.add(edge.dest);
        }
        System.out.println("Kruskal's MST (HashMap):");
        List<Edge> mst2 = kruskalMST_HashMap(verticesSet, new ArrayList<>(edges));
        for (Edge edge : mst2) {
            System.out.println(edge);
        }
        System.out.println("\n--------------------------------------------------------");

        // 3. Kruskal's Algorithm with Path Compression and Rank Optimization
        System.out.println("Kruskal's MST (Optimized):");
        List<Edge> mst3 = kruskalMST_Optimized(numVertices, new ArrayList<>(edges));
        for (Edge edge : mst3) {
            System.out.println(edge);
        }
        System.out.println("\n--------------------------------------------------------");

        // 4. Kruskal's Algorithm for Disconnected Graphs
        System.out.println("Kruskal's MST (Disconnected Graph):");
        List<Edge> mst4 = kruskalMST_Disconnected(numVertices, new ArrayList<>(edges));
        for (Edge edge : mst4) {
            System.out.println(edge);
        }
        System.out.println("\n--------------------------------------------------------");

        // 5. Kruskal's Algorithm with Edge Class Enhancement (Real-World)
        int numNetworkVertices = 6;
        List<NetworkEdge> networkEdges = new ArrayList<>();
        networkEdges.add(new NetworkEdge(0, 1, 100, 10)); // Example: Bandwidth 100, Cost 10
        networkEdges.add(new NetworkEdge(0, 2, 150, 15));
        networkEdges.add(new NetworkEdge(1, 2, 60, 5));
        networkEdges.add(new NetworkEdge(1, 3, 80, 8));
        networkEdges.add(new NetworkEdge(2, 3, 70, 7));
        networkEdges.add(new NetworkEdge(2, 4, 120, 12));
        networkEdges.add(new NetworkEdge(3, 4, 90, 9));
        networkEdges.add(new NetworkEdge(3, 5, 110, 11));
        networkEdges.add(new NetworkEdge(4, 5, 100, 10));

        System.out.println("Kruskal's MST (Network):");
        List<NetworkEdge> mst5 = kruskalMST_Network(numNetworkVertices, networkEdges);
        for (NetworkEdge edge : mst5) {
            System.out.println(edge);
        }
        System.out.println("\n--------------------------------------------------------");
    }
}

